utils/src/main/java/software/amazon/awssdk/utils/uri/SdkUri.java

```diff
@@ -0,0 +1,294 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.utils.uri;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Objects;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.utils.Lazy;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.cache.lru.LruCache;
+import software.amazon.awssdk.utils.uri.internal.UriConstructorArgs;
+
+/**
+ * Global cache for account-id based URI. Prevent calling new URI constructor for the same string, which can cause performance
+ * issues with some uri pattern. Do not directly depend on this class, it will be removed in the future.
+ */
+@SdkProtectedApi
+public final class SdkUri {
+    private static final Logger log = Logger.loggerFor(SdkUri.class);
+
+    private static final String HTTPS_PREFIX = "https://";
+    private static final String HTTP_PREFIX = "http://";
+
+    /*
+     * The default LRUCache size is 100, but for a single service call we cache at least 3 different URIs so the cache size is
+     * increased a bit to account for the different URIs.
+     */
+    private static final int CACHE_SIZE = 150;
+
+    private static final Lazy<SdkUri> INSTANCE = new Lazy<>(SdkUri::new);
+
+    private final LruCache<UriConstructorArgs, URI> cache;
+
+    private SdkUri() {
+        this.cache = LruCache.builder(UriConstructorArgs::newInstance)
+                            .maxSize(CACHE_SIZE)
+                            .build();
+    }
+
+    public static SdkUri getInstance() {
+        return INSTANCE.getValue();
+    }
+
+    public URI create(String s) {
+        if (!isAccountIdUri(s)) {
+            log.trace(() -> "skipping cache for uri " + s);
+            return URI.create(s);
+        }
+        StringConstructorArgs key = new StringConstructorArgs(s);
+        boolean containsK = cache.containsKey(key);
+        URI uri = cache.get(key);
```

Comment:
This isn't 100% reliable since the key could have been inserted between 64 and 65. Do we actually need to log this? If there is, should be do it within `LruCache` instead?

Reply:
we don't actually _need_ to log it, but thought it could help debug if push come to shove.

Reply:
Okay in that case this should be fine, as long as we're not relying on it to be completely accuraet.